### PR TITLE
Add ppl tool

### PR DESCRIPTION
**Pull request by xinyual**
### Description
Add ppl tool and corresponding UT
 
### Issues Resolved
Add ppl tool and corresponding UT
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by ylwu-amzn** [comment]
Github CI failed, fix it first?

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/tools/PPLTool.java [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+                String line = "- " + key + ": " + fieldsToType.get(key);
+                tableInfoJoiner.add(line);
+            }
+        }
+
+        String tableInfo = tableInfoJoiner.toString();
+        return tableInfo;
+    }
+
+    private String constructPrompt(String tableInfo, String question, String indexName)
```
Comment (arjunkumargiri) [inline ml-algorithms/src/main/java/org/opensearch/ml/engine/tools/PPLTool.java:246]: Prompt should not be hard coded in tool but rather be passed as tool spec. This provides flexibility to modify prompt and allow the tool to operate with different remote models.

File: plugin/src/main/java/org/opensearch/ml/autoredeploy/MLModelAutoReDeployer.java [hunk 2]
```diff
@@ -47,6 +47,7 @@
 import org.opensearch.search.sort.SortBuilders;
 import org.opensearch.search.sort.SortOrder;
 
+import lombok.Setter;
```
Comment (arjunkumargiri) [inline plugin/src/main/java/org/opensearch/ml/autoredeploy/MLModelAutoReDeployer.java:50]: Not required.

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/tools/PPLTool.java [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+
+    private String modelId;
+
+    public PPLTool(Client client, String modelId) {
+        this.client = client;
+        this.modelId = modelId;
+    }
+
+    @Override
+    public <T> void run(Map<String, String> parameters, ActionListener<T> listener) {
```
[the remark this answers is not in the document]
Comment (xinyual) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/tools/PPLTool.java:87]: There are some steps like getting samples and corresponding mappings which may only used inside the PPL Tool, I don't think we should divide them into a single tool. Besides, more tools and more interaction between tools will need more calls to LLM, leading to more cost and more mistakes. And according to discussion with Yaliang, current implementation is temporally ok. Maybe we will separate them in the future.

File: plugin/src/main/java/org/opensearch/ml/autoredeploy/MLModelAutoReDeployer.java (hunk 2 above)
Comment (xinyual) [reply plugin/src/main/java/org/opensearch/ml/autoredeploy/MLModelAutoReDeployer.java]: Already removed it.

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/tools/PPLTool.java (hunk 1 above)
Comment (xinyual) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/tools/PPLTool.java:247]: Already add a parameter called contextPrompt. But PPL tool needs a prompt template containing position for question, index and table information. We need to have a detailed description on that so our customer can register it correctly.

**Review by arjunkumargiri — COMMENTED** [review]
* CI is failing, please fix failure

* Can you create a sample agent making use of PPL tool and attach agent execution request/response to PR

**Comment by xinyual** [comment]
> * CI is failing, please fix failure
> 
>     * Can you create a sample agent making use of PPL tool and attach agent execution request/response to PR

I tried a flow agent to execute my tool. The remote model is my finetuned PPL model on sagemaker. To call bedrock, we need a very large prompt. Please refer to langchain POC.

`
POST /_plugins/_ml/agents/_register
{
  "name": "Test_Agent_For_PPL",
  "type": "flow",
  "description": "this is a test agent",
  "tools": [
    {
      "type": "PPLTool",
      "name": "PPLTool",
      "description": "A general tool to generate PPL",
      "parameters": {
        "model_id": "Ue7QGYwB5MH2CK67TMF-",
        "prompt": "Below is an instruction that describes a task, paired with the index and corresponding fields that provides further context. Write a response that appropriately completes the request.\n\n### Instruction:\nI have an opensearch index with fields in the following. Now I have a question: %s Can you help me generate a PPL for that?\n\n### Index:\n%s\n\n### Fields:\n%s\n\n### Response:\n"
      }
    }
  ]
}
`

Then execute the agent with index and question:
`
/_plugins/_ml/agents/<agent_id>/_execute
{
    "parameters": {
        "question": "How many employee with the age greater than 56?",
        "index": "employee"
    }
}
`

Get response like 
`
{
    "inference_results": [
        {
            "output": [
                {
                    "name": "response",
                    "result":"{\"ppl\":\"source\=employee3 | where age \> 56 | stats COUNT() AS count\",\"executionResult\":\"{\\n  \\\"schema\\\": [\\n    {\\n      \\\"name\\\": \\\"count\\\",\\n      \\\"type\\\": \\\"integer\\\"\\n    }\\n  ],\\n  \\\"datarows\\\": [\\n    [\\n      0\\n    ]\\n  ],\\n  \\\"total\\\": 1,\\n  \\\"size\\\": 1\\n}\"}"
                }
            ]
        }
    ]
}
`
The results will contain the ppl and execution result. 

For CI failure,  we still need to wait for PR in the sql: https://github.com/opensearch-project/sql/pull/2438

**Comment by chenryn** [comment]
@xinyual Hello, is there any publicly available documentations about PPL generated model fine-tuning?

**Comment by xinyual** [comment]
> @xinyual Hello, is there any publicly available documentations about PPL generated model fine-tuning?

Currently not. Are you interested in the model weight or the method we fine-tune model?

**Comment by chenryn** [comment]
> > @xinyual Hello, is there any publicly available documentations about PPL generated model fine-tuning?
> 
> Currently not. Are you interested in the model weight or the way we fine-tune model?

I've tried to do something similar myself, but the finetune model doesn't work very well. The model can generate some plausible PPL syntax, but even the differences in questioning habits between me and my colleagues are too great. This make me wonder if this approach works. So it would be nice to have an article about your work (dataset design, etc.).

**Pull request closed without merging**